### PR TITLE
Refactor analyzer

### PR DESCRIPTION
**Pull request by matthewmcnew**

### CODE REVIEW

File: analyzer.go [hunk 1]
```diff
@@ -41,116 +42,186 @@ func (a *Analyzer) Analyze(image image.Image, launchDir string) error {
...
+		return noCacheAvailable
 	}
-	for _, dir := range bpDirs {
-		info, err := os.Stat(dir)
-		if err != nil {
-			return nil, err
+
+	buildpackMetadata, ok := appImageMetadata(abd.groupBP, abd.metaData)
+	if !ok {
+		if cachedToml.Launch == false {
```
Comment (sclevine) [inline analyzer.go]: Use `!cachedTOML.Launch`

File: analyzer.go [hunk 2]
```diff
@@ -41,116 +42,186 @@ func (a *Analyzer) Analyze(image image.Image, launchDir string) error {
...
+	setOfLayers := make(map[string]interface{})
+	buildpackMetadata, ok := appImageMetadata(abd.groupBP, abd.metaData)
+	if ok {
+		for layer := range buildpackMetadata.Layers {
+			setOfLayers[layer] = struct{}{}
+		}
+	}
+
+	bpDir := filepath.Join(abd.launchDir, abd.groupBP)
+	layerTomls, err := filepath.Glob(filepath.Join(bpDir, "*.toml"))
```
Comment (sclevine) [inline analyzer.go]: `layerTOMLs`

File: analyzer.go [hunk 3]
```diff
@@ -179,6 +250,42 @@ func (a *Analyzer) buildpacks() map[string]struct{} {
...
+	bpDirs, err := filepath.Glob(filepath.Join(launchDir, "*"))
+	if err != nil {
+		return nil, err
+	}
+	for _, dir := range bpDirs {
+		info, err := os.Stat(dir)
+		if err != nil {
+			return nil, err
+		}
+		if filepath.Base(dir) != "app" && info.IsDir() {
```
Comment (sclevine) [inline analyzer.go]: Use `os.SameFile` with chosen app dir, as in: https://github.com/buildpack/lifecycle/blob/master/launcher.go

File: analyzer.go [hunk 4]
```diff
@@ -41,116 +42,186 @@ func (a *Analyzer) Analyze(image image.Image, launchDir string) error {
...
 	}
+
 	return nil
 }
 
-func cachedBuildpacks(launchDir string) ([]string, error) {
-	cachedBps := make([]string, 0, 0)
-	bpDirs, err := filepath.Glob(filepath.Join(launchDir, "*"))
+type analyzedBuildPackDirectory struct {
+	metaData  AppImageMetadata
```
Comment (sclevine) [inline analyzer.go]: `metadata` is one word

**Comment by ekcasey** [comment]
@sclevine we incorporated your feedback

**Pull request merged**